qiskit/circuit/library/standard_gates/u.py

```diff
@@ -128,6 +128,35 @@ def __array__(self, dtype=complex):
         )
 
 
+class _CUGateParams(list):
+    # This awful class is to let `CUGate.params` have its keys settable (as
+    # `QuantumCircuit.assign_parameters` requires), while accounting for the problem that `CUGate`
+    # was defined to have a different number of parameters to its `base_gate`, which breaks
+    # `ControlledGate`'s assumptions, and would make most parametric `CUGate`s invalid.
+    #
+    # It's constructed only as part of the `CUGate.params` getter, and given that the general
+    # circuit model assumes that that's a directly mutable list that _must_ be kept in sync with the
+    # gate's requirements, we don't need this to support arbitrary mutation, just enough for
+    # `QuantumCircuit.assign_parameters` to work.
+
+    __slots__ = ("_gate",)
+
+    def __init__(self, gate):
+        super().__init__(gate._params)
+        self._gate = gate
+
+    def __setitem__(self, key, value):
```

Comment:
I think unfortunately there's a non-trivial amount of Qiskit code that assumes that this is a subclass of `list` (like concatenating lists, etc), though I may be confusing that with the old `qargs` and `cargs` entries of the `QuantumCircuit.data` 3-tuple.

Yeah, _technically_ this doesn't prevent all mutation; the full set of things that we'd have to override is: `append`, `clear`, `extend`, `insert`, `pop`, `remove`, `reverse`, `sort`, `__setitem__`, `__delitem__`, `__iadd__` and `__imul__`.  I didn't do that though, because anybody who calls any mutation other than `__setitem__` on an `Instruction.params` already puts it in an invalid state that our existing checking wouldn't catch.  I can put in formal catches for all those if you prefer, though in practice, they'd only affect this class and everything like `RZGate` would be susceptible to the same problems.

(I have the list of overrides immediately to hand, because I needed to do this properly for singletons with parameters!)